src/stk/databases/mongo_db/constructed_molecule.py

```diff
@@ -689,38 +691,78 @@ def _get_molecule_keys(self, entry):
                 yield key, value
 
     def get_all(self):
-        for entry in self._constructed_molecules.find():
-            # Do 'or' query over all key value pairs.
-            query = {'$or': [
-                {key: value}
-                for key, value in self._get_molecule_keys(entry)
-            ]}
-
-            molecule_json = self._molecules.find_one(query)
-            if molecule_json is None:
-                raise KeyError(
-                    'No molecule found in the database associated '
-                    f'with a position matrix with query: {query}. '
-                    'This suggests your database is corrupted.'
-                )
-
-            position_matrix = self._position_matrices.find_one(query)
-            if position_matrix is None:
-                raise KeyError(
-                    'No position matrix found in the database '
-                    'associated with a position matrix with query: '
-                    f'{query}. This suggests your database is '
-                    'corrupted.'
-                )
-
-            yield self._dejsonizer.from_json(
-                json={
-                    'molecule': molecule_json,
-                    'constructedMolecule': entry,
-                    'matrix': position_matrix,
-                    'buildingBlocks': tuple(map(
-                        self._get_building_block,
-                        entry['BB'],
-                    ))
+        # Get all potential indices.
+        indices = chain(
+            self._position_matrices.index_information().values(),
+            self._molecules.index_information().values(),
+            self._constructed_molecules.index_information().values(),
+        )
+        keys = dedupe(index['key'][0][0] for index in indices)
+
+        # Iterate over potential keys, and aggregate matching position
+        # matrices with molecules and constructedmolecules.
+        for key in keys:
+            cursor = self._constructed_molecules.aggregate([
+                {
+                    '$match': {
+                        key: {
+                            '$exists': True,
+                        },
+                    },
                 },
-            )
+                {
+                    '$lookup': {
+                        'from': self._molecules.name,
+                        'localField': key,
+                        'foreignField': key,
+                        'as': 'mol',
+                    },
+                },
+                {
+                    '$lookup': {
+                        'from': self._position_matrices.name,
+                        'localField': key,
+                        'foreignField': key,
+                        'as': 'posmat',
+                    },
+                },
+                {
+                    '$match': {
+                        '$expr': {
+                            '$or': [
```

Comment:
I think this should be "and" not "or"? cos both the mol and the posmat need to exist.